libsystem: Improved support for Windows filesystem

peter-b opened the pull request:
Various useful things for making LCB file syntax more helpful on Windows, mostly relating to LiveCode to-and-from native path conversion.  Deals with lots of weird corner cases more nicely.

livecodefraser commented on libfoundation/src/system-file-w32.cpp:
```diff
@@ -416,17 +545,16 @@ __MCSFilePathFromNative (MCStringRef p_native_path,
 		return false;
 
-	for (uindex_t i = 0; i < t_len; ++i)
+	for (uindex_t i = 0; i < t_len, ++i)
```
I suspect this change was an accident!

peter-b commented:
@livecodefraser Pull request updated.

livecodefraser commented:
@livecode-vulcan review okay fe41c4d

livecode-vulcan commented:
💙 review by @livecodefraser ok fe41c4d98b1b64fd0c03398464837e07a63662d3

livecode-vulcan commented:
😞 test failure fe41c4d98b1b64fd0c03398464837e07a63662d3
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1805)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1818)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1799)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1811)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1821)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1824)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1830)

peter-b commented:
I think it's safe to assume @livecodefraser is okay with adding a semicolon.

@livecode-vulcan review by @livecodefraser ok 6398d7b

livecode-vulcan commented:
💙 review by @livecodefraser ok 6398d7b94271192b8c6029f83675e879a49e85a1

livecode-vulcan commented:
😞 test failure 6398d7b94271192b8c6029f83675e879a49e85a1
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1807)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1820)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1801)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1813)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1823)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1826)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1832)

peter-b commented:
@livecode-vulcan retry 6398d7b

livecode-vulcan commented:
😞 test failure 6398d7b94271192b8c6029f83675e879a49e85a1

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/100)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/100)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/86)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/32)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/32)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/98)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/100)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/102)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/104)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/102)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/104)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/120)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/101)

peter-b committed:
libsystem: Add API for retrieving CWD.

Add `MCSFileGetCurrentDirectory()` for retrieving the process's
current working directory.

peter-b committed:
libsystem: Add MCSFilePathIsAbsolute()

Add a way to test whether a path is absolute or not.

peter-b committed:
[emscripten] Use MCSFileGetCurrentDirectory() in GetCurrentFolder().

Remove lots of unnecessary code.

peter-b committed:
libsystem: Tweak Windows path handling.

* Internal->native translation:
  * `\` characters are invalid
  * "Device" files (e.g. "NUL", "COM1") are invalid

* Native->internal translation:
  * `\` and `/` both map to `/`
  * "Device" files (e.g. "NUL", "COM1") are invalid

runrevmark commented:
@livecode-vulcan review ok aaf52f5

livecode-vulcan commented:
💙 review by @runrevmark ok aaf52f5232102a0fb161f6acc1245fe0b63df217

livecode-vulcan commented:
😎 test success aaf52f5232102a0fb161f6acc1245fe0b63df217

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/146)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/148)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/134)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/78)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/78)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/144)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/148)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/148)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/152)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/148)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/152)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/172)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/145)
